### PR TITLE
Create spec.md

### PR DESCRIPTION
**Pull request by zpatrick**

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# ORCA
+
+## EKS Limitations
+* EKS currently does not allow users to configure their cluster's [authentication](https://docs.aws.amazon.com/eks/latest/userguide/managing-auth.html) format.
+This prevents us from using a [proxy](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy) or
+[webhook](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) service.
+* Kubernetes does have a concept of users, but they [cannot be added through an API call in Kubernetes](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes), 
+since it is assumed that users are "...to be managed by an outside, independent service". 
+Users that are managed by these types of systems (e.g. Auth0) can only be created by using a proxy or webhook service (NOTE: this is an assumption - verification is still needed!).
```
Comment (zpatrick) [inline docs/spec.md]: All of the **NOTE** sections will be deleted; I want to use them to  to start points of discussion.

File: docs/spec.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+#### Option 2
+3.5 After initially creating the user's corresponding service account, the **orca** webservice does some permission management on behalf of the cluster admin.
+Some options:
+* Create a role binding from the service account to some existing role. 
+* Create a new role and role binding for the service account.
+* Add the service account to a group. 
+* Create a namespace with the same name as the service account, create an **admin** role for the namespace, and create a role binding from the service account to the namespace's **admin** role. 
+(NOTE: It is 
+
+NOTE: Option 1 is probably more desireable in the open-source frame-of-mind since the **orca** tool isn't bound to EKS,
```
Comment (zpatrick) [inline docs/spec.md]: Would like to open discussion here on which option preferred. Personally, I'm going with "configurable".

Comment (zpatrick) [inline docs/spec.md]: Would like to open discussion here on which option is preferred. Personally, I'm going with "configurable".

File: docs/spec.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+## Authentication Flow
+This workflow assumes there is an already-running EKS cluster with the Orca web service running in it. 
+The Orca web service will need permissions to manage service accounts and namespaces in the cluster. 
+The user needs to generate a `kubeconfig` file to authenticate themselves via Auth0 with that cluster. 
+
+1. The user navigates to the **orca** webservice endpoint and clicks on the `Generate Kubeconfig` link. 
+2. That link redirects the user to Auth0 to perform the OAuth2 handshake.
+3. After a successful handshake:
+  * If a service account for the corresponding user does not exist: the **orca** webservice creates a new service account in the **default** or **kube-system** [namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#working-with-namespaces)
```
Comment (zpatrick) [inline docs/spec.md]: From the link:
> * default The default namespace for objects with no other namespace
> * kube-system The namespace for objects created by the Kubernetes system

**default** seems more appropriate to me given those descriptions since "we", not the "Kubernetes system", create the objects. Maybe there are caveats I'm not seeing?

File: docs/spec.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+It is the responsibility of the cluster administrator to ensure the permissions of the newly-created service account are properly set.
+This depends on which [Authorization Module](https://kubernetes.io/docs/reference/access-authn-authz/authorization/#authorization-modules) the cluster is setup to use. 
+
+
+#### Option 2
+3.5 After initially creating the user's corresponding service account, the **orca** webservice does some permission management on behalf of the cluster admin.
+Some options:
+* Create a role binding from the service account to some existing role. 
+* Create a new role and role binding for the service account.
+* Add the service account to a group.
```
Comment (zpatrick) [inline docs/spec.md]: [nit] "Add the service account to some existing group"

File: docs/spec.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+#### Option 2
+3.5 After initially creating the user's corresponding service account, the **orca** webservice does some permission management on behalf of the cluster admin.
+Some options:
+* Create a role binding from the service account to some existing role. 
+* Create a new role and role binding for the service account.
+* Add the service account to a group. 
+* Create a namespace with the same name as the service account, create an **admin** role for the namespace, and create a role binding from the service account to the namespace's **admin** role. 
+(NOTE: It is
```
Comment (zpatrick) [inline docs/spec.md]: oops

File: docs/spec.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+#### Option 2
+3.5 After initially creating the user's corresponding service account, the **orca** webservice does some permission management on behalf of the cluster admin.
+Some options:
+* Create a role binding from the service account to some existing role. 
+* Create a new role and role binding for the service account.
+* Add the service account to some existing group. 
+* Create a namespace with the same name as the service account, create an **admin** role for the namespace, and create a role binding from the service account to the namespace's **admin** role. 
+
+NOTE: Option 1 is probably more desireable in the open-source frame-of-mind since the **orca** tool isn't bound to EKS,
```
Comment (zpatrick) [inline docs/spec.md]: Would like to open discussion here on which option is preferred. Personally, I'm going with "configurable".

File: docs/spec.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+The Orca web service will need permissions to manage service accounts and namespaces in the cluster. 
+The user needs to generate a `kubeconfig` file to authenticate themselves via Auth0 with that cluster. 
+
+1. The user navigates to the **orca** webservice endpoint and clicks on the `Generate Kubeconfig` link. 
+2. That link redirects the user to Auth0 to perform the OAuth2 handshake.
+3. After a successful handshake:
+  * If a service account for the corresponding user does not exist: the **orca** webservice creates a new service account in the **default** or **kube-system** [namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#working-with-namespaces)
+(NOTE: the **default** namespace may be more appropriate than **kube-system**) that corresponds to that user. 
+  * If a service account for the corresponding user already exists: the **orca** webservice will lookup the existing service account's authentication token. 
+  * The webservice generates a `Kubeconfig` using the service account's authentication token.
```
Comment (zpatrick) [inline docs/spec.md]: This will work if service account tokens do not expire. If they do, then we would need a binary that can call out to the orca web service to grab a new token.

File: docs/spec.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
+# ORCA
+
+## EKS Limitations
```
Comment (diemonster) [inline docs/spec.md]: [nit] consider renaming this section 'BACKGROUND'

File: docs/spec.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+* EKS currently does not allow users to configure their cluster's [authentication](https://docs.aws.amazon.com/eks/latest/userguide/managing-auth.html) format.
+This prevents us from using a [proxy](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy) or
+[webhook](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) service.
+* Kubernetes does have a concept of users, but they [cannot be added through an API call in Kubernetes](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes), 
+since it is assumed that users are "...to be managed by an outside, independent service". 
+Users that are managed by these types of systems (e.g. Auth0) can only be created by using a proxy or webhook service (NOTE: this is an assumption - verification is still needed!).
```
Comment (diemonster) [inline docs/spec.md]: I'm not sure if this line is true: "Users [...] can only be created by using a proxy or webhook service". Or at the very least, I'm unclear on what it's trying to say as 'proxy' in this context is a bit ambiguous.

From the k8s docs:
> Normal users are assumed to be managed by an outside, independent service. An admin distributing private keys, a user store like Keystone or Google Accounts, even a file with a list of usernames and passwords. In this regard, Kubernetes does not have objects which represent normal user accounts. Normal users cannot be added to a cluster through an API call.

ref: https://kubernetes.io/docs/reference/access-authn-authz/authentication/ (this is a doc that is worth several re-reads)

File: docs/spec.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+[webhook](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) service.
+* Kubernetes does have a concept of users, but they [cannot be added through an API call in Kubernetes](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes), 
+since it is assumed that users are "...to be managed by an outside, independent service". 
+Users that are managed by these types of systems (e.g. Auth0) can only be created by using a proxy or webhook service (NOTE: this is an assumption - verification is still needed!). 
+Therefore, our concept of users will be mapped to [service accounts](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes) since those can be managed via the Kubernetes API,
+and are functionally very similar to users. 
+
+
+## Authentication Flow
+This workflow assumes there is an already-running EKS cluster with the Orca web service running in it.
```
Comment (diemonster) [inline docs/spec.md]: [nit] ~This workflow~ The proposed workflow

Comment (diemonster) [inline docs/spec.md]: Consider adding an ABSTRACT section to the beginning of the doc which, in one sentence, explains what the Orca web service is. Orca hasn't been introduced until this moment in the doc. 

e.g. https://github.com/golang/proposal/blob/master/design/12166-subtests.md

File: docs/spec.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+## Authentication Flow
+This workflow assumes there is an already-running EKS cluster with the Orca web service running in it. 
+The Orca web service will need permissions to manage service accounts and namespaces in the cluster. 
+The user needs to generate a `kubeconfig` file to authenticate themselves via Auth0 with that cluster. 
+
+1. The user navigates to the **orca** webservice endpoint and clicks on the `Generate Kubeconfig` link. 
+2. That link redirects the user to Auth0 to perform the OAuth2 handshake.
+3. After a successful handshake:
+  * If a service account for the corresponding user does not exist: the **orca** webservice creates a new service account in the **default** or **kube-system** [namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#working-with-namespaces)
```
Comment (diemonster) [inline docs/spec.md]: why would the **default** namespace be more appropriate?

File: docs/spec.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+This prevents us from using a [proxy](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy) or
+[webhook](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) service.
+* Kubernetes does have a concept of users, but they [cannot be added through an API call in Kubernetes](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes), 
+since it is assumed that users are "...to be managed by an outside, independent service". 
+Users that are managed by these types of systems (e.g. Auth0) can only be created by using a proxy or webhook service (NOTE: this is an assumption - verification is still needed!). 
+Therefore, our concept of users will be mapped to [service accounts](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes) since those can be managed via the Kubernetes API,
+and are functionally very similar to users. 
+
+
+## Authentication Flow
```
Comment (diemonster) [inline docs/spec.md]: I think there needs to be more detail on what exactly this spec is concerned with --- if it's strictly authentication flow, then we should call that out.

File: docs/spec.md (hunk 6 above)
Comment (diemonster) [reply docs/spec.md]: At the moment, this spec is discounting the usefulness of allowing (or aiding) one-to-many style group => role relationships. We know that the auth request to the k8s API has a field which can specify `group` membership, so what's the argument for ignoring this functionality? To repeat the answer to your question from yesterday re: the difference between groups and roles:

> A group is a collection of users with a given set of permissions assigned to the group (and transitively, to the users). A role is a collection of permissions, and a user effectively inherits those permissions when he acts under that role. 

I'm not sure why this concept is only addressed in option #2. This idea is something that is useful to anyone; it's a simple idea that is present is most (all?) other RBAC implementations I can think of.

ref: https://stackoverflow.com/questions/7770728/group-vs-role-any-real-difference

File: docs/spec.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
+# ORCA
+
+## EKS Limitations
+* EKS currently does not allow users to configure their cluster's [authentication](https://docs.aws.amazon.com/eks/latest/userguide/managing-auth.html) format.
+This prevents us from using a [proxy](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy) or
+[webhook](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) service.
+* Kubernetes does have a concept of users, but they [cannot be added through an API call in Kubernetes](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes), 
+since it is assumed that users are "...to be managed by an outside, independent service". 
+Users that are managed by these types of systems (e.g. Auth0) can only be created by using a proxy or webhook service (NOTE: this is an assumption - verification is still needed!). 
+Therefore, our concept of users will be mapped to [service accounts](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes) since those can be managed via the Kubernetes API,
```
Comment (diemonster) [inline docs/spec.md]: Sometimes I think this might just be a bad idea altogether. Yes it will work, but is it the right thing to do?

> User accounts are for humans. Service accounts are for processes, which run in pods.

ref: https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/

File: docs/spec.md (hunk 3 above)
Comment (nehayward) [reply docs/spec.md]: Yeah I think default or create our own, `orca-system`

File: docs/spec.md (hunk 6 above)
Comment (zpatrick) [reply docs/spec.md]: I'm assuming you wrote this before you go to [here](https://github.com/quintilesims/orca/pull/1/files#diff-6e196c98aaadaa54d9ba9bbe92a13320R39)?

File: docs/spec.md (hunk 13 above)
Comment (diemonster) [reply docs/spec.md]: If we were to keep in line with what Kubernetes seems to want you to do, then Orca would be a delegation service. Users wouldn't get their own token, but rather they would auth to Orca and Orca itself would make the request to the k8s API on their behalf.

In other words, Orca would be a middle manager that has one or many service accounts that users essentially "use" per request. This way we're not strictly tying users to service accounts & tokens, but rather users to some organizational unit (groups? lol) within Orca, which then maps to a service of service accounts that has role(s) assigned to them based off permission needs.

File: docs/spec.md (hunk 6 above)
Comment (zpatrick) [reply docs/spec.md]: > At the moment, this spec is discounting the usefulness of allowing (or aiding) one-to-many style group => role relationship.

I disagree, in fact there's a suggestion for using groups right [here](https://github.com/quintilesims/orca/pull/1/files#diff-6e196c98aaadaa54d9ba9bbe92a13320R39). 

The distinction between these two options is the degree to which the service performs administrative tasks on behalf of the cluster admin after a new service account is created. 
Option 1 is to not do any tasks and leave it up to the cluster admin.
Option 2 is to do some tasks on behalf of the cluster admin. The [Some options](https://github.com/quintilesims/orca/pull/1/files#diff-6e196c98aaadaa54d9ba9bbe92a13320R36) list is just a brainstorming of what those tasks may be.

Comment (zpatrick) [reply docs/spec.md]: Perhaps I haven't been clear enough about an underlying problem I've been trying to address, so let me try to do that here:

K8 lets you choose one of many access control methods: RBAC, ABAC, Node, etc. 
For each of those given methods, users implement their own permission models. 
Whenever we add a requirement to our service that requires a specific access control method, or a certain permission model for that access control method, we reduce the portability of our service. 

If we design this correctly, each of the following people should be able to use our service without changing their permission model:
* Someone wants to use ABAC in the cluster
* Someone wants to use RBAC with only roles in their permission model. 
* Someone wants to use RBAC with a combination of roles and groups in their permission model. 

TLDR: Our service needs to fit into other peoples' permission models, not the other way around.
